pageindicatorview/src/main/java/com/rd/PageIndicatorView2.java

```diff
@@ -0,0 +1,882 @@
+    }
+
+    /**
+     * Set progress value in range [0 - 1] to specify state of animation while selecting new circle indicator.
+     *
+     * @param selectingPosition selecting position with specific progress value.
+     * @param progress          float value of progress.
+     */
+    public void setProgress(int selectingPosition, float progress) {
+        Indicator indicator = manager.indicator();
+        if (!indicator.isInteractiveAnimation()) {
+            return;
+        }
+
+        int count = indicator.getCount();
+        if (count <= 0 || selectingPosition < 0) {
+            selectingPosition = 0;
+
+        } else if (selectingPosition > count - 1) {
+            selectingPosition = count - 1;
+        }
+
+        if (progress < 0) {
+            progress = 0;
+
+        } else if (progress > 1) {
+            progress = 1;
+        }
+
+        if (progress == 1) {
+            indicator.setLastSelectedPosition(indicator.getSelectedPosition());
+            indicator.setSelectedPosition(selectingPosition);
+        }
+
+        indicator.setSelectingPosition(selectingPosition);
+        manager.animate().interactive(progress);
+    }
+
+    public void setClickListener(@Nullable DrawController.ClickListener listener) {
+        manager.drawer().setClickListener(listener);
+    }
+
+    private void init(@Nullable AttributeSet attrs) {
+        setupId();
+        initIndicatorManager(attrs);
+
+        if (manager.indicator().isFadeOnIdle()) {
+            startIdleRunnable();
+        }
+
+        pageChangeCallback = new ViewPager2.OnPageChangeCallback() {
+            @Override
+            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
+                onPageScroll(position, positionOffset);
+            }
+
+            @Override
+            public void onPageScrollStateChanged(int state) {
+                if (state == ViewPager2.SCROLL_STATE_IDLE) {
+                    manager.indicator().setInteractiveAnimation(isInteractionEnabled);
+                }
+            }
+        };
+    }
+
+    private void setupId() {
+        if (getId() == NO_ID) {
+            setId(IdUtils.generateViewId());
+        }
+    }
+
+    private void initIndicatorManager(@Nullable AttributeSet attrs) {
+        manager = new IndicatorManager(this);
+        manager.drawer().initAttributes(getContext(), attrs);
+
+        Indicator indicator = manager.indicator();
+        indicator.setPaddingLeft(getPaddingLeft());
+        indicator.setPaddingTop(getPaddingTop());
+        indicator.setPaddingRight(getPaddingRight());
+        indicator.setPaddingBottom(getPaddingBottom());
+        isInteractionEnabled = indicator.isInteractiveAnimation();
+    }
+
+    private void registerSetObserver() {
+        if (setObserver != null || viewPager == null || viewPager.getAdapter() == null) {
+            return;
+        }
+
+        setObserver = new RecyclerView.AdapterDataObserver() {
```

Comment:
Should have other overrides to work with Android's `AdapterListUpdateCallback`. 
`override fun onItemRangeInserted(positionStart: Int, itemCount: Int)` is important because without it, there are cases where the page indicator view won't display when it should. Same with implementing `override fun onItemRangeChanged(positionStart: Int, itemCount: Int)`.